datastream-kafka/src/main/java/com/linkedin/datastream/kafka/KafkaProducerWrapper.java

```diff
@@ -254,67 +298,103 @@ void send(DatastreamTask task, ProducerRecord<K, V> producerRecord, Callback onC
         }
       } catch (Exception e) {
         _log.error(String.format("Send failed for partition %d with an exception", producerRecord.partition()), e);
-        throw generateSendFailure(e);
+        throw generateSendFailure(e, task);
       }
     }
   }
 
   @VisibleForTesting
   void shutdownProducer() {
     Producer<K, V> producer;
-    synchronized (_producerLock) {
+    try {
+      _producerLock.lock();
+      // if there is no producer or the producer close is in progress, return.
+      if (_kafkaProducer == null || _closeInProgress) {
+        return;
+      }
       producer = _kafkaProducer;
+      _closeInProgress = true;
       // Nullify first to prevent subsequent send() to use
```

Comment:
is this comment still relevant now that producer access is protected with lock? The code is not harmful, but may be the comment can go if this is not relevant as it might cause some confusion on the intent of the code.

Reply:
Actually I feel the comment helps in understanding the intent, even though we have made sure that the producer close in progress will block the send anyway. In the send() if the producer is not null, it does not check for close in progress.